### DISCUSSION TITLE
Validations are 1 step behind

### DISCUSSION THREAD

**Discussion by LeyliAbbas** [react-hook-form/react-hook-form#5073] [Q&A]
I have used `react-hook-forms` + yup scheme for form validations. But, validations don't work properly. I try to debug `console.log(this.options?.context)` and see that the context value in yup scheme shows the old value. So, validations are 1 step behind. How can I resolve this problem?

useForm() hook

```
const {
    handleSubmit,
    control,
    reset,
    formState: { errors },
    trigger,
    watch
  } = useForm({
    mode: 'all',
    reValidateMode: 'onChange',
    criteriaMode: 'firstError',
    context: {
      fromAccountBlockers: context?.fromAccountBlockers
    },
    resolver: yupResolver(ownPaymentSchema)
  });
```

Form element

```
  <Controller
    control={control}
    name="accountFrom"
    defaultValue={undefined}
    render={({ field }) => (
      <SelectAccount
        label={t('account')}
        account={field.value}
        setAccount={(data, option) => {
          field.onChange(data);
          setContext({
            ...context,
            fromAccountBlockers: option.blockers
          });
        }}
        error={!!errors.accountFrom}
        message={errors.accountFrom && t(errors.accountFrom.message)}
      />
    )}
  />
```

Yup scheme

```
const ownPaymentSchema = yup.object().shape({
  accountFrom: yup
    .string()
    .required('errorMessages.commissionAccountRequired')
    .test('accountHasBlockers', 'errorMessages.accountHasBlockers', function () {
      console.log(this.options?.context)
      return this.options?.context?.fromAccountBlockers?.length === 0;
    })
});
```

**Comment by bluebill1049** [comment]
can you try to reproduce that in a codesandbox?

**Reply by LeyliAbbas** [reply]
https://codesandbox.io/s/react-hook-form-yup-resolver-huj0w - You can see similar problem here

**Comment by bluebill1049** [comment]
<img width="453" alt="Screen Shot 2021-05-07 at 11 05 43 pm" src="https://user-images.githubusercontent.com/10513364/117453846-c1658100-af88-11eb-80ef-582af254b124.png">

defaultValue can't be `undefined`, and you not sending anything to hook form `field.onChange();`

**Reply by bluebill1049** [reply]
try defaultValue="" and field.onChange(e)

**Reply by LeyliAbbas** [reply]
I did as you said but the same problem persists :(

**Reply by bluebill1049** [reply]
oh, I miss your `context`, you are using useState, it will be 1 render behind because onChange will trigger a re-render and flush the context, you may need to use useRef for that context value, but I think it's better to avoid context if you can supply value through the onChange.
